### PR TITLE
Backport: Gutenberg 44731

### PR DESCRIPTION
**Pull request by scruffian**
This backports https://github.com/WordPress/gutenberg/pull/44731

I couldn't get it to work locally because the CSS file didn't seem to be there, but maybe I'm testing it wrong?

### CODE REVIEW

**Comment by ockham** [comment]
> I couldn't get it to work locally because the CSS file didn't seem to be there, but maybe I'm testing it wrong?

I haven't tested it yet, but did you `npm i` before building to make sure you've got the latest package versions?

**Comment by scruffian** [comment]
Yes, but I still didn't see the CSS file...

**Comment by ockham** [comment]
Hmm, I _am_ getting 

```
  | <br />
-- | --
  | <b>Warning</b>:  file_get_contents(/wp-includes/css/dist/block-library/classic.css): failed to open stream: No such file or directory in <b>/var/www/src/wp-includes/script-loader.php</b> on line <b>3696</b><br />
  | <br />
  | <b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /var/www/src/wp-includes/script-loader.php:3696) in <b>/var/www/src/wp-admin/admin-header.php</b> on line <b>9</b><br />
```

File: src/wp-includes/script-loader.php
Comment (ockham) [inline src/wp-includes/script-loader.php]: I think it's the leading slash. This might work fine for "smart" functions such as `wp_register_style`, but `file_get_contents` will take this literally -- for the file system root 😱

Comment (ockham) [reply src/wp-includes/script-loader.php]: This seems to do the trick:

```diff
diff --git a/src/wp-includes/script-loader.php b/src/wp-includes/script-loader.php
index 8f9e420d3c..fd91718ebf 100644
--- a/src/wp-includes/script-loader.php
+++ b/src/wp-includes/script-loader.php
@@ -3687,13 +3687,13 @@ function wp_add_editor_classic_theme_styles( $editor_settings ) {
                return $editor_settings;
        }
        $suffix = wp_scripts_get_suffix();
-       $classic_theme_styles = "/wp-includes/css/dist/block-library/classic$suffix.css";
+       $classic_theme_styles = "/css/dist/block-library/classic$suffix.css";
 
        // This follows the pattern of get_block_editor_theme_styles,
        // but we can't use get_block_editor_theme_styles directly as it
        // only handles external files or theme files.
        $editor_settings['styles'][] = array(
-               'css'            => file_get_contents( $classic_theme_styles ),
+               'css'            => file_get_contents( __DIR__ . $classic_theme_styles ),
                'baseURL'        => get_theme_file_uri( $classic_theme_styles ),
                '__unstableType' => 'theme',
                'isGlobalStyles' => false,
```

I don't know about [`get_theme_file_uri`](https://developer.wordpress.org/reference/functions/get_theme_file_uri/). Can that ever work as it's supposed to for a file inside `wp-includes`?

Comment (ockham) [reply src/wp-includes/script-loader.php]: Maybe we don't need the `baseURL` field?

Comment (ockham) [reply src/wp-includes/script-loader.php]: Fixed the path in d1b37eb33e9fc3e746a8e09cad78426c8e1d62b3, and removed `baseURL` in 1bc847445b2decc4948ba6fb8ae149bc4bed0b97.

File: src/wp-includes/script-loader.php
Comment (ockham) [inline src/wp-includes/script-loader.php]: Maybe this should be `core`? It's arguably Core that's providing this styling 🤔

Comment (ockham) [reply src/wp-includes/script-loader.php]: cc/ @oandregal @jorgefilipecosta for Global Styles expertise 😊

Comment (ockham) [reply src/wp-includes/script-loader.php]: I'll tentatively change it to `core` since IMO it better represents where the styling is coming from. 

Since this is targeting classic themes, it shouldn't have any actual implications, since this is for Global Styles -- which are unavailable for Classic Themes.

**Review by ockham — DISMISSED** [review]
Thanks! This LGTM and fixes the [issue](https://github.com/WordPress/gutenberg/issues/44659#issuecomment-1268305786) for me:

In the post editor, with Twenty Ten active:

image

**Comment by ockham** [comment]
Ugh, unit tests are failing because they cannot locate that file:

```
file_get_contents(/var/www/src/wp-includes/css/dist/block-library/classic.css): failed to open stream: No such file or directory
```

I'll look into this tomorrow. Not sure if something is off with the way we're building things for unit tests, or if there's another reason 😕

File: src/wp-includes/script-loader.php
Comment (scruffian) [reply src/wp-includes/script-loader.php]: Thanks, that's a good idea.

File: src/wp-includes/script-loader.php
Comment (scruffian) [reply src/wp-includes/script-loader.php]: Thank you

**Comment by scruffian** [comment]
We found another bug with the original solution, which I have a fix for here: https://github.com/WordPress/gutenberg/pull/45063

I'll also backport that in this PR

**Comment by ockham** [comment]
> Ugh, unit tests are failing because they cannot locate that file:
> 
> ```
> file_get_contents(/var/www/src/wp-includes/css/dist/block-library/classic.css): failed to open stream: No such file or directory
> ```
> 
> I'll look into this tomorrow. Not sure if something is off with the way we're building things for unit tests, or if there's another reason 😕

Some quick thoughts on this:
- My current working hypothesis is that maybe we're not building `@wordpress` package JS and CSS before running PHP unit tests.
- We have other functions (e.g. `wp_default_styles`) that register styles that might (?) not exist at the time. How do they do it? (Or is it all classic stylesheets are readily available at the time that tests are run?)
- Are there any helpers we can use instead of `file_get_contents` that will avoid the "not found" error at test time (via some indirection?)
- Finally, do we need to mock that file, or `file_get_contents` when run for that file, or all of `wp_add_editor_classic_theme_styles`?

**Comment by scruffian** [comment]
Would it be simpler to just use a static CSS file?

**Comment by ockham** [comment]
> Would it be simpler to just use a static CSS file?

Probably 🤔 Maybe due to my ignorance -- there _might_ be established patterns to deal with built CSS in a scenario like this, I'm just unaware of it.

Would we lose anything significant if we made it static?

**Comment by scruffian** [comment]
> Would we lose anything significant if we made it static?

Nothing at all

**Comment by scruffian** [comment]
Done in 20fed67, lets hope the tests pass!

**Comment by cbravobernal** [comment]
Using TwentyTwenty Twelve, it seems that the font color of the buttons is different 😢 
The CSS specificity seems to be the same on all themes, I tried with 2010, 2011, 2012, and 2013, so maybe that error is not 100% related to this backport.
Also, the theme CSS file is loading correctly in all of them.

Editor:
Screenshot 2022-10-19 at 14 10 19
Frontend:
Screenshot 2022-10-19 at 14 10 23

**Comment by scruffian** [comment]
I added a commit to fix Twenty Twelve

**Review by ockham — APPROVED** [review]
Thank you @scruffian! LGTM :shipit:

File: src/wp-includes/script-loader.php
Comment (oandregal) [reply src/wp-includes/script-loader.php]: :wave: Looked at where `__unstableType` is used. It's only at https://github.com/WordPress/gutenberg/blob/88a1f7ae18bea3a13f6d1e507e3c25090b76cb08/packages/edit-post/src/editor.js#L165 

If I'm reading the original [PR correctly](https://github.com/WordPress/gutenberg/pull/35736#discussion_r733460913), this is used to decide whether to render a particular stylesheet when the user deactivates "theme styles". If the stylesheet needs to be disabled by that check, we should use "theme", otherwise "core" is fine. See check in this screen:

imagem

Comment (scruffian) [reply src/wp-includes/script-loader.php]: These styles are the "default" ones set for buttons by core, they aren't connected to a specific theme, so I think `core` is correct.

**Review by audrasjb — APPROVED** [review]
_[committer double signoff]_
The changeset looks good to me as well. 

Self assigning for commit.

**Edit: Not yet. See the comment below.**

**Review by audrasjb — CHANGES_REQUESTED** [review]
Hmm, I think the PR is missing RTL styles, at least for Twenty Twenty.

**Comment by sabernhardt** [comment]
Yes, Twenty Twenty also needs the RTL stylesheet, but block styles would belong in [editor-style-block.css](https://github.com/WordPress/wordpress-develop/blob/8ed4c1099755fdda8615d1cb4164c593b3a99031/src/wp-content/themes/twentytwenty/assets/css/editor-style-block.css#L925) instead of classic.

Twenty Twelve does not have an RTL stylesheet for the editor.

**Comment by scruffian** [comment]
> Yes, Twenty Twenty also needs the RTL stylesheet, but block styles would belong in [editor-style-block.css](https://github.com/WordPress/wordpress-develop/blob/8ed4c1099755fdda8615d1cb4164c593b3a99031/src/wp-content/themes/twentytwenty/assets/css/editor-style-block.css#L925) instead of classic.

The styles are already in `editor-style-block.css`, but the problem is that the editor block styles aren't provided through `add_editor_style` as they are usually. This means they don't get prefixed with `editor-styles-wrapper`, so they don't have the same specificity as the ones coming from this new CSS file. An alternative would be to manually add this prefix to the rules inside `editor-style-block.css` which I am happy to do if you think that approach is better.

**Comment by sabernhardt** [comment]
The new inline styles come after `editor-style-block.css` with equal specificity. If you make the selectors slightly more specific for [lines 927-928](https://github.com/WordPress/wordpress-develop/blob/8ed4c1099755fdda8615d1cb4164c593b3a99031/src/wp-content/themes/twentytwenty/assets/css/editor-style-block.css#L927-L928) and the font size adjustment at [lines 1308-1309](https://github.com/WordPress/wordpress-develop/blob/8ed4c1099755fdda8615d1cb4164c593b3a99031/src/wp-content/themes/twentytwenty/assets/css/editor-style-block.css#L1308-1309), that should be enough to correct the original state without changing the theme's Classic Editor stylesheet.

```
:root .editor-styles-wrapper .wp-block-button__link,
:root .editor-styles-wrapper .wp-block-file__button {
```

Twenty Twenty does not change text color on hover, but you could add the underline in `editor-style-block.css`, without `:root`. (Button block elements are not focusable in the editor.)

```
.editor-styles-wrapper .wp-block-button__link:hover,
.editor-styles-wrapper .wp-block-file__button:hover {
	text-decoration: underline;
}
```

**Comment by scruffian** [comment]
@sabernhardt thanks that's super helpful. I have updated editor-style-blocks and revert the changes to the classic styles. I also updated the RTL stylesheet. Thanks again for all your help.

**Review by ockham — APPROVED** [review]
I retested Twenty Twenty in Hebrew to verify that RTL styling works ✅ 

image

I also tested Twenty Ten and Twenty Twelve to verify that button styling still works there ✅ 

Changes look good to me 👍

**Review by audrasjb — APPROVED** [review]

**Comment by audrasjb** [comment]
_[committer double signoff]_

Yeah thanks for the changes. Looks good to me now 👍

**Comment by ockham** [comment]
Commit message draft:

```
Editor: Correctly apply Button block styles for classic themes

In [54358], styling for Button blocks that had been removed from classic themes was reintroduced. However, it was added with a global scope, whereas editor styles are usually added with a `.editor-styles-wrapper` selector, which makes them more specific.

This change modifies the way that classic theme styles are added so that they also get wrapped in an `.editor-styles-wrapper` selector to match specificity.

Furthermore, adjust specificity for some Button block related styling in the editor for the Twenty Twelve and Twenty Twenty themes.

Merges [https://github.com/WordPress/gutenberg/pull/44731 Gutenberg PR 44731] into trunk.

Follow-up to [54358].
Props scruffian, cbravobernal, sabernhardt, audrasjb.
See #56467.

```

**Comment by ockham** [comment]
Committed to Core `trunk` in [r54687](https://core.trac.wordpress.org/changeset/54687). [Awaiting second approval](https://core.trac.wordpress.org/ticket/56467#comment:387) to backport to the `6.1` branch.

**Comment by ockham** [comment]
> Committed to Core `trunk` in [r54687](https://core.trac.wordpress.org/changeset/54687). [Awaiting second approval](https://core.trac.wordpress.org/ticket/56467#comment:387) to backport to the `6.1` branch.

Merged to the `6.1` branch in [r54688](https://core.trac.wordpress.org/changeset/54688).

**Pull request closed without merging**
